File: CudaOpenCV/Python/GPUtest/gpu_test.py
# https://learnopencv.com/getting-started-opencv-cuda-module/

import sys
import cv2

if len(sys.argv) < 2:
    print(f"Program usage: {sys.argv[0]} image_filename")
    exit(1)

img = cv2.imread(sys.argv[1], cv2.IMREAD_GRAYSCALE)

src = cv2.cuda_GpuMat()
src.upload(img)

clahe = cv2.cuda.createCLAHE(clipLimit=5.0, tileGridSize=(8, 8))
dst = clahe.apply(src, cv2.cuda_Stream.Null())

result = dst.download()

cv2.imshow("result", result)
cv2.waitKey(0)

File: CudaOpenCV/Cpp/GPUtest/gpu_test.cpp
#include <iostream>
#include <opencv2/cudaimgproc.hpp>
#include <opencv2/highgui.hpp>

// https://learnopencv.com/getting-started-opencv-cuda-modul/

int main(int argc, char* argv[]) {
  if (argc < 2) {
    std::cout << "Program usage: " << argv[0] << " image_filename\n";
    return -1;
  }

  cv::Mat img = cv::imread(argv[1], cv::IMREAD_GRAYSCALE);
  cv::cuda::GpuMat dst;
  cv::cuda::GpuMat src;
  src.upload(img);

  cv::Ptr<cv::cuda::CLAHE> ptr_clahe =
      cv::cuda::createCLAHE(5.0, cv::Size(8, 8));
  ptr_clahe->apply(src, dst);

  cv::Mat result;
  dst.download(result);

  cv::imshow("result", result);
  cv::waitKey();

  return 0;
}
